Storage/src/Connection/RetryTrait.php

```diff
@@ -0,0 +1,293 @@
+     * @param array $args arguments which include preconditions provided,
+     *  eg: ['ifGenerationMatch' => 0].
+     * @return bool
+     */
+    private function isPreConditionSupplied($methodName, array $args)
+    {
+        if (isset(self::$condIdempotentOps[$methodName])) {
+            // return true if required precondition are given.
+            return !empty(array_intersect(
+                self::$condIdempotentOps[$methodName],
+                array_keys($args)
+            ));
+        }
+        return false;
+    }
+
+    /**
+     * Decide whether the op needs to be retried or not.
+     *
+     * @param \Exception $exception The exception object received
+     * while sending the request.
+     * @param int $currentAttempt Current retry attempt.
+     * @param bool $isIdempotent
+     * @param bool $preconditionNeeded
+     * @param bool $preconditionSupplied
+     * @param int $maxRetries
+     * @return bool
+     */
+    private function retryDeciderFunction(
+        \Exception $exception,
+        $currentAttempt,
+        $isIdempotent,
+        $preconditionNeeded,
+        $preconditionSupplied,
+        $maxRetries,
+        $retryStrategy
+    ) {
+        if ($retryStrategy == self::$RETRY_STRATEGY_NEVER) {
+            return false;
+        }
+
+        $statusCode = $exception->getCode();
+        // Retry if the exception status code matches
+        // with one of the retriable status code and
+        // the operation is either idempotent or conditionally
+        // idempotent with preconditions supplied.
+
+        if (in_array($statusCode, self::$httpRetryCodes)) {
+            if ($retryStrategy == self::$RETRY_STRATEGY_ALWAYS) {
+                return true;
+            } elseif ($isIdempotent) {
+                return true;
+            } elseif ($preconditionNeeded) {
+                return $preconditionSupplied;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * Getter function that returns the RETRY_STRATEGY_ALWAYS key.
+     */
+    public static function getStrategyAlwaysKey()
+    {
+        return self::$RETRY_STRATEGY_ALWAYS;
+    }
+
+    /**
+     * Getter function that returns the RETRY_STRATEGY_NEVER key.
+     */
+    public static function getStrategyNeverKey()
+    {
+        return self::$RETRY_STRATEGY_NEVER;
+    }
+
+    /**
+     * Getter function that returns the RETRY_STRATEGY_IDEMPOTENT key.
+     */
+    public static function getStrategyIdempotentKey()
+    {
+        return self::$RETRY_STRATEGY_IDEMPOTENT;
+    }
+
+    /**
+     * Getter function that returns the HTTP retry codes.
+     */
+    public static function getHttpRetryCodes()
+    {
+        return self::$httpRetryCodes;
+    }
+
+    /**
+     * Utility func that returns the list of headers that need to be
+     * attached to every request and its retries.
+     */
+    public static function getRetryHeaders($invocationId, $attemptCount)
```

Comment:
We do not want to add public methods if we don't need to. There's no reason to add these. See https://github.com/googleapis/google-cloud-php/pull/6075/files

Reply:
I think was needed when we had the download retry code in `StorageObject`, with that latest change, we can make it private.